server/cli/upgrade.py

```diff
@@ -0,0 +1,82 @@
+import click
+import re
+import requests
+
+from requests.exceptions import ConnectionError
+from .. import __version__
+
+# Official SemVer regex: https://semver.org/
+SEMVER_FORMAT = re.compile(
+    r"^(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)"
+    + r"(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)"
+    + r"(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?"
+    + r"(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$"
+)
+
+
+def log_upgrade_check():
+    # Sanity-check that the CLI version is a properly-formatted SemVer string
+    assert validate_version_str(__version__, release_only=False)
+
+    # Get the current latest release
+    try:
+        release_tag_generator = (r['tag_name'] for r in _request_cellxgene_releases())
+        latest_release = next(release_tag_generator, lambda tag_name: validate_version_str(tag_name))
+        if version_gt(latest_release, __version__):
+            click.echo(f"There's a new version of cellxgene available ({latest_release})!")
+            click.echo("To upgrade, run the following: pip install --upgrade cellxgene\n")
+    except (ConnectionError, RateLimitException):
+        click.echo("Upgrade check failed.\n")
+
+
+class RateLimitException(Exception):
+    """
+    Github API Rate Limit Exception
+    """
+
+
+def _request_cellxgene_releases():
+    def raise_on_rate_limit(response):
+        if response.status_code == 403:
+            raise RateLimitException
+    url = "https://api.github.com/repos/chanzuckerberg/cellxgene/releases"
+    res = requests.get(url)
+    for release in res.json():
+        yield release
+    while 'next' in res.links.keys():
+        res = requests.get(res.links['next']['url'])
+        for release in res.json():
+            yield release
+
+
+def validate_version_str(version_str, release_only=True):
+    """
+    Test if a string conforms to SemVer format (https://semver.org/)
```

Comment:
The regex can also throw an AttributeError, right?  Should it return False in that case or let the exception propagate?

Reply:
and if the exception is part of the API, add a test

Reply:
I've rewritten the function such that the call that would cause the `AttributeError` (`match.group("prerelease")`) will only be called in the case that there _is_ a match. In this case, no `AttributeError` can be raised.